### PR TITLE
Adding new mode to StereoSGBM and introducing HAL acceleration

### PR DESCRIPTION
**Pull request by sbokov**
Current speedup is around 2-3x with almost the same quality as MODE_SGBM.

### CODE REVIEW

File: modules/calib3d/src/stereosgbm.cpp [hunk 1]
```diff
@@ -829,6 +822,636 @@ static void computeDisparitySGBM( const Mat& img1, const Mat& img2,
...
+                for( d = 0; d < D; d++ )
+                {
+                    if( horPassCostVolume[x+d]*(100 - uniquenessRatio) < min_cost*100 && std::abs(d - best_d) > 1 )
+                        break;
+                }
+#endif
+                if( d < D )
+                    continue;
+            }
+            d = best_d;
```
Comment (mshabunin) [inline modules/calib3d/src/stereosgbm.cpp]: The ARM builder produces a warning: 
`stereosgbm.cpp:1383:39: warning: ‘best_d’ may be used uninitialized in this function [-Wmaybe-uninitialized]`

**Comment by sbokov** [comment]
@mshabunin Fixed. Could you rerun the ARM builder to verify?

**Comment by sbokov** [comment]
@mshabunin I changed the StereoSGBM ROI to make it similar to StereoBM. However, there is a bit of quality degradation due to the fact that the left-right consistency check now has less information to validate disparity values: https://www.dropbox.com/sh/v2wnt1gmujoruy6/AAAFZGlq0NufCb9BGTT6bt2la?dl=0 . It makes the algorithm a bit faster though, as it now doesn't compute disparity values for pixels, belonging to the left border. If that's ok, we will also need to update the tests.

File: modules/calib3d/src/stereosgbm.cpp [hunk 2]
```diff
@@ -829,6 +822,645 @@ static void computeDisparitySGBM( const Mat& img1, const Mat& img2,
...
+    curCostVolumeLine  = (CostType*)alignPtr(buffer.ptr(), 16);
+    hsumBuf            = curCostVolumeLine + costVolumeLineSize;
+    pixDiff            = hsumBuf + costVolumeLineSize*hsumBufNRows;
+    tmpBuf             = (PixType*)(pixDiff + costVolumeLineSize);
+    horPassCostVolume  = (CostType*)(tmpBuf + width*16*num_ch);
+    vertPassCostVolume = horPassCostVolume + costVolumeLineSize_ext;
+    rightPassBuf       = vertPassCostVolume + costVolumeLineSize_ext;
+    vertPassMin        = rightPassBuf + D;
+    disp2CostBuf       = vertPassMin + width1_ext;
+    disp2Buf           = disp2CostBuf + width;
```
Comment (sbokov) [inline modules/calib3d/src/stereosgbm.cpp]: @mshabunin I tried replacing this part with separate 16-byte-aligned memory allocations for each buffer. It still worked fine for me, i.e. no invalid memory accesses.

**Comment by mshabunin** [comment]
Great! Thank you!
:+1:

**Pull request merged**